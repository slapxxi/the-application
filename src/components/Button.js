// @flow
import _ from 'lodash';
import styled from 'react-emotion';

const Button = styled('button')`
  padding: ${15 / 1.5}px 15px;
  border: 0;
  border-radius: 2px;
  text-transform: uppercase;
  font-size: 14px;
  font-weight: bold;
  background: ${(p) => _.get(p, 'theme.colors.button')};
  color: ${(p) => _.get(p, 'theme.colors.buttonText')};
  cursor: pointer;

  &:active {
    transform: translateY(1px);
    background: ${(p) => _.get(p, 'theme.colors.buttonActive')};
    color: ${(p) => _.get(p, 'theme.colors.buttonTextActive')};
  }

  &:disabled {
    background: lightgrey;
    color: #bbbbbb;
    cursor: not-allowed;
  }
`;

export default Button;
