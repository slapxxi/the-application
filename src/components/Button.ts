import { get } from 'lodash';
import { darken, invert, lighten } from 'polished';
import { px } from '../lib';
import styled from 'styled-components';

const Button = styled.button`
  border-radius: 4px;
  font-weight: 500;
  line-height: 0;
  vertical-align: middle;
  cursor: pointer;

  &:disabled {
    background: 'lightgrey';
    color: #bbb;
    cursor: not-allowed;
  }

  ${({ theme }) => {
    const font = get(theme, 'sizes.text');
    const padding = get(theme, 'sizes.padding', 10) / 1.5;
    const background = get(theme, 'colors.button', 'black');
    return {
      border: `1px solid ${background}`,
      color: invert(background),
      fill: invert(background),
      fontSize: px(font),
      padding: px(padding),
      '&:hover': {
        background,
        color: lighten(0.9, invert(background)),
      },
      '&:active': {
        borderColor: darken(0.1, background),
        background: darken(0.1, background),
        color: lighten(0.1, invert(background)),
      },
    };
  }};
`;

export default Button;
