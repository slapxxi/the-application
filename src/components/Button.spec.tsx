import * as React from 'react';
import { render } from 'react-testing-library';
import Button from './Button';

const theme = {
  colors: {
    button: 'black',
    buttonActive: 'white',
    buttonText: 'white',
    buttonTextActive: 'black',
  },
};

it('renders', () => {
  const { container } = render(<Button>button</Button>);
  expect(container.firstChild).toMatchSnapshot();
});

it('accepts `disabled` prop', () => {
  const { container } = render(<Button disabled>button</Button>);
  expect(container.firstChild).toMatchSnapshot();
});

it('accepts `theme` prop', () => {
  const { container } = render(<Button theme={theme}>button</Button>);
  expect(container.firstChild).toMatchSnapshot();
});

it('does not throw if `theme` is malformed', () => {
  expect(() =>
    render(<Button theme={{}}>button</Button>),
  ).not.toThrow();
});
