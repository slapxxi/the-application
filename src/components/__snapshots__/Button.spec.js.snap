// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`accepts \`disabled\` prop 1`] = `
.emotion-0 {
  padding: 10px 15px;
  border: 0;
  border-radius: 2px;
  text-transform: uppercase;
  font-size: 14px;
  font-weight: bold;
  cursor: pointer;
}

.emotion-0:active {
  -webkit-transform: translateY(1px);
  -ms-transform: translateY(1px);
  transform: translateY(1px);
}

.emotion-0:disabled {
  background: lightgrey;
  color: #bbbbbb;
  cursor: not-allowed;
}

<button
  class="emotion-0 emotion-1"
  disabled=""
>
  button
</button>
`;

exports[`accepts \`theme\` prop 1`] = `
.emotion-0 {
  padding: 10px 15px;
  border: 0;
  border-radius: 2px;
  text-transform: uppercase;
  font-size: 14px;
  font-weight: bold;
  background: black;
  color: white;
  cursor: pointer;
}

.emotion-0:active {
  -webkit-transform: translateY(1px);
  -ms-transform: translateY(1px);
  transform: translateY(1px);
  background: white;
  color: black;
}

.emotion-0:disabled {
  background: lightgrey;
  color: #bbbbbb;
  cursor: not-allowed;
}

<button
  class="emotion-0 emotion-1"
>
  button
</button>
`;

exports[`renders 1`] = `
.emotion-0 {
  padding: 10px 15px;
  border: 0;
  border-radius: 2px;
  text-transform: uppercase;
  font-size: 14px;
  font-weight: bold;
  cursor: pointer;
}

.emotion-0:active {
  -webkit-transform: translateY(1px);
  -ms-transform: translateY(1px);
  transform: translateY(1px);
}

.emotion-0:disabled {
  background: lightgrey;
  color: #bbbbbb;
  cursor: not-allowed;
}

<button
  class="emotion-0 emotion-1"
>
  button
</button>
`;
